tests/js/server/replication2/replication2-maintenance-replicated-state-cluster.js

```diff
@@ -0,0 +1,184 @@
+    SH.updateReplicatedStatePlan(database, logId, function () {
+      let log = {
+        id: logId,
+        currentTerm: {
+          term: 1,
+          config: {
+            replicationFactor: 3,
+            writeConcern: 2,
+            softWriteConcern: 2,
+            waitForSync: false,
+          },
+          leader: {
+            serverId: leader,
+            rebootId: LH.getServerRebootId(leader),
+          }
+        }, targetConfig: {
+          replicationFactor: 3,
+          writeConcern: 2,
+          softWriteConcern: 2,
+          waitForSync: false,
+        },
+        participantsConfig: {
+          generation: 1,
+          participants: {},
+        }
+      };
+      let state = {
+        id: logId,
+        generation: 1,
+        participants: {},
+        properties: {
+          implementation: {
+            type: "black-hole",
+          }
+        }
+      };
+      for (const server of servers) {
+        state.participants[server] = {
+          generation: 1,
+        };
+        log.participantsConfig.participants[server] = {};
+      }
+      return {state, log};
+    });
+  };
+
+  return {
+    setUpAll, tearDownAll,
+    setUp: LH.registerAgencyTestBegin,
+    tearDown: LH.registerAgencyTestEnd,
+
+    testCreateReplicatedState: function () {
+      const logId = LH.nextUniqueLogId();
+      const servers = _.sampleSize(LH.dbservers, 3);
+      const leader = servers[0];
+      createReplicatedState(database, logId, servers, leader);
+      LH.waitFor(spreds.replicatedStateIsReady(database, logId, servers));
+    },
+
+    testReplicatedStateUpdateParticipantGeneration: function () {
+      const logId = LH.nextUniqueLogId();
+      const servers = _.sampleSize(LH.dbservers, 3);
+      const leader = servers[0];
+      createReplicatedState(database, logId, servers, leader);
+
+      LH.waitFor(spreds.replicatedStateIsReady(database, logId, servers));
+
+
+      const followers = _.difference(servers, [leader]);
+      const follower = _.sample(followers);
+      SH.updateReplicatedStatePlan(database, logId, function (state, log) {
+        state.participants[follower].generation += 1;
+        return {state, log};
+      });
+
+      LH.waitFor(spreds.replicatedStateIsReady(database, logId, servers));
+    },
+
+    testReplicatedStateChangeLeader: function () {
+      const logId = LH.nextUniqueLogId();
+      const servers = _.sampleSize(LH.dbservers, 3);
+      const leader = servers[0];
+      createReplicatedState(database, logId, servers, leader);
+
+      LH.waitFor(spreds.replicatedStateIsReady(database, logId, servers));
+
+      SH.updateReplicatedStatePlan(database, logId, function (state, log) {
+        log.currentTerm.term += 1;
+        return {state, log};
+      });
+
+      LH.waitFor(LH.replicatedLogIsReady(database, logId, 2, servers, leader));
+      LH.waitFor(spreds.replicatedStateIsReady(database, logId, servers));
```

Comment:
As only the log was explicitly changed, wouldn't this last assertion be true even if the replicated state didn't notice anything? Can we even check anything else?

Reply:
What would you expect to happen from a replicated state point of view?

Reply:
Nothing in particular, but from a test point of view, we cannot observe anything.

Maybe it makes sense, after inserting a participant into the state, to report the term of that participant in current/state?

Reply:
Since this would require an additional feature I would like to do this in a separate PR.

I will create a [ticket](https://arangodb.atlassian.net/browse/CINFRA-242) for that.